torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -334,20 +338,25 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
 
   // Try moving persistent buffer factors into threads until we have too many
   // threads.
+  constexpr int batches_per_block_inner_reduction_max = 10;
   while (
-      // If using less than a quarter of available threads
-      bdimx * bdimy * bdimz * 2 <=
-          ceilDiv(device_max_threads_per_multiprocessor, (int64_t)4) &&
+      // If block size can be doubled
+      bdimx * bdimy * bdimz * 2 <= max_threads_in_block &&
       // And batches_per_block_inner_reduction can be divided by two
-      (batches_per_block_inner_reduction >= 2 ||
+      (batches_per_block_inner_reduction >=
+           batches_per_block_inner_reduction_max ||
        batches_per_block_outer_reduction >= 2)) {
     // Try to decrease per thread register allocation persistence size on inner
     // reduction
-    if (batches_per_block_inner_reduction >= 2 &&
-        batches_per_block_inner_reduction !=
-            roundUpPow2Or8(batches_per_block_inner_reduction / 2)) {
-      batches_per_block_inner_reduction =
-          roundUpPow2Or8(batches_per_block_inner_reduction / 2);
+    if (batches_per_block_inner_reduction >=
+        batches_per_block_inner_reduction_max) {
+      if (batches_per_block_inner_reduction * bdimx ==
+          threads_after_vectorize) {
+        batches_per_block_inner_reduction /= 2;
+      } else {
+        batches_per_block_inner_reduction =
```

Comment:
If can't be uniformly distributed, we want to keep a relatively large PB by just minus one. e.g. e.g. case 10272, `threads_after_vectorize= 10272/8=1284`, `batches_per_block_inner_reduction=11`, and `bdimx=117`. After this piece of code, `batches_per_block_inner_reduction=9`, and `bdimx=143`. Otherwise, `batches_per_block_inner_reduction=5` . Bandwidth will drop from 990 GB/s to 970 GB/s. Not significant, I can actually simplify this part to avoid "overfit".
Simplified version should be:
```
    if (batches_per_block_inner_reduction >=
        batches_per_block_inner_reduction_max) {
      batches_per_block_inner_reduction /= 2;
      bdimx = ceilDiv(
          inner_most_dimension_numel,
          inner_reduction_unroll_factor * batches_per_block_inner_reduction);
      continue;
    }
```
Prefer original or simplified?

Reply:
I'm feeling it's becoming too specific to particular problems on particular devices. Non-divisible problem sizes are tricky to develop robust heuristics. I'd suggest it as an open problem and focus on relatively easier sizes in this PR.

Reply:
changed to simplified version.